iocore/eventsystem/UnixEThread.cc

```diff
@@ -142,6 +152,141 @@ EThread::process_event(Event *e, int calling_code)
+  ink_hrtime next_time = 0;
+  ink_hrtime delta     = 0;    // time spent in the event loop
+  ink_hrtime loop_start_time;  // Time the loop started.
+  ink_hrtime loop_finish_time; // Time at the end of the loop.
+
+  // Track this so we can update on boundary crossing.
+  EventMetrics *prev_metric = this->prev(metrics + (ink_get_hrtime_internal() / HRTIME_SECOND) % N_EVENT_METRICS);
+
+  int nq_count = 0;
+  int ev_count = 0;
+
+  // A statically initialized instance we can use as a prototype for initializing other instances.
+  static EventMetrics METRIC_INIT;
+
+  // give priority to immediate events
+  for (;;) {
+    if (unlikely(shutdown_event_system == true)) {
+      return;
+    }
+
+    // execute all the available external events that have
+    // already been dequeued
+    loop_start_time = cur_time = Thread::get_hrtime_updated();
+    nq_count                   = 0; // count # of elements put on negative queue.
+    ev_count                   = 0; // # of events handled.
+
+    current_metric = metrics + (loop_start_time / HRTIME_SECOND) % N_EVENT_METRICS;
+    if (current_metric != prev_metric) {
+      // Mixed feelings - really this shouldn't be needed, but just in case more than one entry is
+      // skipped, clear them all.
+      do {
+        memcpy((prev_metric = this->next(prev_metric)), &METRIC_INIT, sizeof(METRIC_INIT));
+      } while (current_metric != prev_metric);
+      current_metric->_loop_time._start = loop_start_time;
+    }
+    ++(current_metric->_count);
+
+    // Move events from the external thread safe queues to the local queue.
+    EventQueueExternal.dequeue_external();
+
+    while ((e = EventQueueExternal.dequeue_local())) {
+      ++ev_count;
+      if (e->cancelled) {
+        free_event(e);
+      } else if (!e->timeout_at) { // IMMEDIATE
+        ink_assert(e->period == 0);
+        process_event(e, e->callback_event);
+      } else if (e->timeout_at > 0) { // INTERVAL
+        EventQueue.enqueue(e, cur_time);
+      } else { // NEGATIVE
+        Event *p = nullptr;
+        Event *a = NegativeQueue.head;
+        while (a && a->timeout_at > e->timeout_at) {
+          p = a;
+          a = a->link.next;
+        }
+        if (!a) {
+          NegativeQueue.enqueue(e);
+        } else {
+          NegativeQueue.insert(e, p);
+        }
+        ++nq_count;
+      }
+    }
+
+    bool done_one;
+    do {
+      done_one = false;
+      // execute all the eligible internal events
+      EventQueue.check_ready(cur_time, this);
+      while ((e = EventQueue.dequeue_ready(cur_time))) {
+        ink_assert(e);
+        ink_assert(e->timeout_at > 0);
+        if (e->cancelled)
+          free_event(e);
+        else {
+          done_one = true;
+          process_event(e, e->callback_event);
+        }
+      }
+    } while (done_one);
+
+    // execute any negative (poll) events
+    if (NegativeQueue.head) {
+      // execute poll events
+      while ((e = NegativeQueue.dequeue())) {
```

Comment:
Before execute the NegativeQueue, we should check the ProtectQueue again because it may be a long time to execute the EventQueue.